### DISCUSSION TITLE
muting a plugin data saving protocol

### DISCUSSION THREAD

**Discussion by gorkemer** [jspsych/jsPsych#1141] [Q&A] [answered]
Hello everyone,

In our experiment, there is one test trial and one feedback trial. I put them in a timeline like this:

```
var testProcedureGain = {
    timeline: [test_trial, feedback_trial],
    timeline_variables: test_images
};
```

This is all good. But when it comes to saving to MySQL database, I see rows of feedback trials which are mostly “null”. This is normal because feedback is actually a trial and jspsych’s plugin defaults run its data-saving procedure. Please see the image attached here.

Is there a common way to entirely stop a plugin trial from saving to the database? Something like a mute?

1- I tried to merge the feedback trial into the test-trial. I can edit the plugin by changing the `after_response` function and use `display_element` and then add var = html_content = “something”, but there I can only use `response.key` and I need not just keyboard response but more from test trial feature parameters, and I can’t seem to access them in there (those results in ‘undefined’). 

This is how I do normally access trial parameters to be used used for feedback response. 

```
on_finish: function (data) {
      	var pair_type = jsPsych.data.get().last(1).values()[0].pair_type; // test stimuli type
	var pair_ori = jsPsych.data.get().last(1).values()[0].pair_ori; // test stimuli orientation
```
..

2- Another way to do that would be doing it after on_finish:
Is there a way to add an extra function to use display_element?
`on_finish: function (data, display_element)`
Show something using display_element
I can’t seem to work this around either.

I also tried removing trial_data from the `jsPsych.finishTrial()`, but I don’t solve the problem. 

Any help in finding a direction is much appreciated! 

Gorkem

<img width="850" alt="Screen Shot 2020-10-25 at 10 10 25 AM" src="https://user-images.githubusercontent.com/16053523/97112589-c7f67000-16aa-11eb-93b9-c54fde3c4775.png">

**Comment by becky-gilbert** [comment, marked as answer]
Hi Gorkem,

Could you just remove all of the feedback trials from the jsPsych data before you send the data back to the server? For instance using jsPsych's data manipulation functions like [filter](https://www.jspsych.org/core_library/jspsych-data/#filter) or [filterCustom](https://www.jspsych.org/core_library/jspsych-data/#filtercustom)?

Otherwise your first option seems like it should work. You can edit the plugin so that it takes the correct response as a parameter (or any other information you need to score the response, like type and orientation).
```javascript
plugin.info = {
  name: 'html-keyboard-response',
  description: '',
  parameters: {

    // other parameters here...

    stim_type: {
      type: jsPsych.plugins.parameterType.STRING,
      pretty_name: 'stim type',
      default: null,
      description: 'Stim type.'
    },
    stim_orientation: {
      type: jsPsych.plugins.parameterType.STRING,
      pretty_name: 'stim orientation',
      default: null,
      description: 'Stim orientation.'
    },
}
```
And then you can pass those values into the plugin when you set up the trial, so you'll be able to use that information in the after_response function to determine what the feedback text should be. You'll also want to change the plugin so that the end_trial function is only called after your feedback display is finished. 
```javascript
var after_response = function(info) {
  // after a valid response, the stimulus will have the CSS class 'responded'
  // which can be used to provide visual feedback that a response was recorded
  display_element.querySelector('#jspsych-html-keyboard-response-stimulus').className += ' responded';

  // only record the first response
  if (response.key == null) {
    response = info;
  }

  // response.key tells you what key was pressed
  // trial.stim_type and trial.stim_orientation give you the info that you need to score the response
  // set up a variable to hold the feedback text, and use the response and condition info
  // to store the appropriate feedback text in this variable
  var feedback;
  if (trial.stim_type == "x" & trial.stim_orientation == "y" & response.key == "a") {
    feedback = "Correct!";
  } else {
    feedback = "Incorrect";
  }
  // display feedback
  display_element.innerHTML = feedback;
  // call the end_trial function after some delay, e.g. 3 seconds
  jsPsych.pluginAPI.setTimeout(function() {
    end_trial();
  }, 3000);
}
```
I don't think your second option will work because by the time the on_finish function is called, the trial is over, and there's really no way to get jsPsych to stop there while you show feedback on the screen (but I could be wrong!).

**Comment by gorkemer** [comment]
@becky-gilbert Thank you very much, Becky. The data manipulation function of filter() solved my problem very quickly, I should have known that but I thank you for the other solution as well. It would also work.

Best,
Gorkem
